### DISCUSSION TITLE
Observable Flutter: rrousel's simple infinite scroll example clarification

### DISCUSSION THREAD

**Discussion by Weav3r** [rrousselGit/riverpod#2887] [Q&A]
I watched the episode [Observable Flutter: Diving into Riverpod](https://www.youtube.com/watch?v=BJtQ0dfI-RA&t=2760s) with Remi. It was a good episode and I learnt a lot. However, rrousel gave a simple infinite scroll example which blew my mind when I first saw it with such minimal code but there are some quirks that I'd like esteemed knowledgeable folks here to help me with. I'm aware there's the marvel example which provides a complete example of infinite scrolling but at this point in time it goes way over my head but I'm still going through it to understand. So here's the minimum reproduction of the example from the episode.

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';

part 'main.g.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return ProviderScope(
      child: MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
          useMaterial3: true,
        ),
        home: const MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends ConsumerStatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  ConsumerState<MyHomePage> createState() => _MyHomePageState();
}

@riverpod
Future<List<String>> fetchItem(FetchItemRef, {required int page}) async {
  if (page > 2) return ['a', 'b'];
  await Future.delayed(Duration(seconds: 3));
  return List.generate(50, (index) => 'Hello $page $index');
}

class _MyHomePageState extends ConsumerState<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: ListView.builder(itemBuilder: ((context, index) {
        final page = index ~/ 50;
        final itemIndex = index % 50;

        final pageValue = ref.watch(fetchItemProvider(page: page));
        return pageValue.when(
          data: (items) {
            if (itemIndex >= items.length) return null;
            return Center(child: Text(items[itemIndex]));
          },
          loading: () {
            //Below line ensures only one loading indicator is returned on scroll down
            // but causes issues with rapid scroll back
            if (itemIndex != 0) return null;
            return Center(child: CircularProgressIndicator());
          },
          error: (err, st) => Text('Error'),
        );
      })),
    );
  }
}
```
The example works well but during the bit he talked about showing only one loading indicator see comment in `loading: ()...`. I noticed that scrolling back after the provider is disposed causes the screen to be blank and takes multiple scroll ups to display the data again. This was even observed in the video. He spoke about using keepAlive to keep the provider from being disposed immediately but for a social media app where users can scroll down for hours and then rapidly scroll back up, keeping the provider alive for hours doesn't seem like a good idea to me (I stand to be corrected). So basically, is there a way to only display one loading indicator on both scroll down and scroll up. Things I've tried include:

```
              if (itemIndex == 0 || itemIndex == 49)
                return Center(child: CircularProgressIndicator());
              return null;
```
and also tried to use scroll direction to determine the index to return the loading indicator but with no luck. Any help would be greatly appreciated. Thank you.

**Comment by AhmedLSayed9** [comment]
>  but for a social media app where users can scroll down for hours and then rapidly scroll back up, keeping the provider alive for hours doesn't seem like a good idea to me (I stand to be corrected)

As far as i know, they keep the state cached when you scroll down, but they limit how far you can scroll.

i.e: Twitter allows you to view only 3200 tweets when scrolling. After you reach the limit, you can go back and get another random list. If you're at someone's profile with too many historical tweets, you can't pass that limit either and you've to use the search or some filtering.

The thing is, the items widgets themselves get dynamically removed and re-added as they scroll out of and back into the view (in addition to cacheExtent if used). but the state of the items can be cached (by riverpod in example) so when you scroll back up, you don't have to load the items again and afaik that's fine.

Furthermore, Instead of using keep-alive which requires you to manually invalidate the family provider when you pop the screen, you can use a better trick: let each page watch/listen to the prev page to keep all page's state cached. This will also allow you to dispose the pages at the end when you scroll back to the top (Another trick: you can delay their disposal so that when the user scroll up and then scroll down fast, he can find the same data. but If he stays up for x minutes, the cached data at the bottom will be cleared).

Finally, if you still need to demand items on scrolling down and up, current `ListView.builder` implementation doesn't support that but there's an open [issue](https://github.com/flutter/flutter/issues/130658) for it.

**Reply by rrousselGit** [reply]
Social medias also often use "cacheExtent" to dispose of items if they are far from view.

**Reply by Weav3r** [reply]
I never change the "cacheExtent" believing the defaults to be good enough. I've seen a couple answers on the web that propose using values such as 999+. @rrousselGit Do you think such values are justified even for a social media app?

**Reply by AhmedLSayed9** [reply]
> Social medias also often use "cacheExtent" to dispose of items if they are far from view.

Isn't `cacheExtent` supposed to do the opposite?
I mean that it early load items that are not yet on screen (to hide loading and provide smooth scrolling)

**Comment by Weav3r** [comment]
@AhmedLSayed9 Thank you for your insightful response. Just a few comment and follow up questions.

> The thing is, the items widgets themselves get dynamically removed and re-added as they scroll out of and back into the view (in addition to cacheExtent if used). but the state of the items can be cached (by riverpod in example) so when you scroll back up, you don't have to load the items again and afaik that's fine.

Thank you, I'm aware that the widgets gets reused and the state can be cached with riverpod. I'm just a bit skeptical about memory use when caching that many states (indirectly objects).

> you can use a better trick: let each page watch/listen to the prev page to keep all page's state cached. This will also allow you to dispose the pages at the end when you scroll back to the top

Genius! Hadn't thought about that.

>(Another trick: you can delay their disposal so that when the user scroll up and then scroll down fast, he can find the same data. but If he stays up for x minutes, the cached data at the bottom will be cleared).

That will be using keep-alive right? That's the only way I know to delay disposal. Please correct me if I'm wrong.

**Reply by AhmedLSayed9** [reply]
> I'm just a bit skeptical about memory use when caching that many states (indirectly objects).

Well, if you're using auto-dispose provider. they will be disposed anyway when you leave the screen (or some minutes after if you're delaying the disposal a bit). I don't know how much memory it will need, but you can test it and decide the maximum scroll limit based on that. I've tested Facebook/Twitter before, scrolled down for minutes and scrolled back to the top, all cached :)

> That will be using keep-alive right? That's the only way I know to delay disposal. Please correct me if I'm wrong.

You can implement it using keepAliveLink + a timer, check https://github.com/rrousselGit/riverpod/issues/1664#issuecomment-1264608305

**Reply by Weav3r** [reply]
Thanks for the clarification.

>You can implement it using keepAliveLink + a timer, check #1664 (comment)

This is exactly what Remi did in the video.

**Comment by rrousselGit** [comment]
Ultimately the issue is that if a previous page disappeared, you don't want to show a single loading indicator but rather a shimmer of the list item, for the full page length.
Otherwise you'll jump in index, which means the index of the current item jumps too, which gets weird.

On top of my head, one solution could be to check if the "next page" exists when loading.

```dart
final page = index / pageLength;
final pageContent = ref.watch(paginatedProvider(page: page));

if (pageContent.isLoading) {
  final hasNextPage = ref.exist(paginatedProvider(page: page + 1));
  final isFirstItemInPage = page % pageLength == 0;
  // Likely scrolling back up, so showing a shimmer of the list item
  if (hasNextPage) return Shimmer();
  // Scrolling down, let's show a single spinner
  if (isFirstItemInPage) return CircularProgressIndicator();
  return null;
}
```

I haven't run this code so who knows, but I think that's something worth investigating
